src/Controller/Component.php

```diff
@@ -170,12 +170,19 @@ public function implementedEvents(): array
             'Controller.startup' => 'startup',
             'Controller.beforeRender' => 'beforeRender',
             'Controller.beforeRedirect' => 'beforeRedirect',
-            'Controller.shutdown' => 'shutdown',
+            'Controller.shutdown' => 'afterFilter',
         ];
         $events = [];
         foreach ($eventMap as $event => $method) {
             if (method_exists($this, $method)) {
                 $events[$event] = $method;
+
+                continue;
+            }
+
+            if ($event === 'Controller.shutdown' && method_exists($this, 'shutdown')) {
```

Comment:
The `$eventMap` is static so checking the event name inside the loop is unnecessary. Can instead do the following after the loop:
```
if (!isset($events['Controller.shutdown']) && method_exists($this, 'shutdown'))` {
   ...
}
```